File: backend/app/extraction/dxf_extractor.py
"""DXF existant/projet extraction.

Two methods, auto-detected (Method B preferred when the DXF declares an
explicit AXE-*/COTE-* layer convention, else Method A, the geometric
heuristic): both ultimately produce Bands + WidthSamples calibrated against
the shared AxisReference built from the mandatory 'Axes + profils' file, so
existant and projet always land on the same PK grid.
"""

from __future__ import annotations

import math
import re
from collections import defaultdict

from app.extraction.axis_reference import AxisReference
from app.extraction.dxf_common import (
    find_legacy_polylines,
    load_document,
    named_axis_and_cote_layers,
    read_texts,
)
from app.extraction.geometry import PolylineIndex, order_lines_by_offset, perpendicular_direction
from app.models.domain import Band, ElementType, Side, SourceMethod, StateKind, WidthSample

ROUTE_TEMPLATE_HALF = [ElementType.ACCOTEMENT, ElementType.TROTTOIR, ElementType.CYCLE, ElementType.VOIE]
AUTOROUTE_TEMPLATE_HALF = [ElementType.ACCOTEMENT, ElementType.BAU, ElementType.VOIE, ElementType.VOIE]

GABARIT_TEMPLATES: dict[str, dict] = {
    "route": {"half": ROUTE_TEMPLATE_HALF, "center": None},
    "autoroute": {"half": AUTOROUTE_TEMPLATE_HALF, "center": ElementType.TPC},
}

_ELEMENT_TOKENS: list[tuple[str, ElementType]] = [
    (r"BAU", ElementType.BAU),
    (r"ACCOT", ElementType.ACCOTEMENT),
    (r"TR.{0,4}OIR", ElementType.TROTTOIR),
    (r"CYCL", ElementType.CYCLE),
    (r"VOIE|CHAUSS", ElementType.VOIE),
    (r"TPC", ElementType.TPC),
]


def detect_mode(doc) -> str:
    return "layers" if named_axis_and_cote_layers(doc) is not None else "heuristic"


def extract_dxf_state(
    path: str,
    state: StateKind,
    axis: AxisReference,
    gabarit: str = "route",
    step_m: float = 5.0,
    band_overrides: dict[str, tuple[Side, ElementType]] | None = None,
) -> tuple[list[Band], list[WidthSample], str]:
    doc = load_document(path)
    mode = detect_mode(doc)
    band_overrides = band_overrides or {}

    if mode == "layers":
        bands, samples = _extract_layer_based(doc, state, axis, band_overrides)
    else:
        bands, samples = _extract_heuristic(doc, state, axis, gabarit, step_m, band_overrides)
    return bands, samples, mode


def _default_band_labels(n_bands: int, gabarit: str) -> list[tuple[Side, ElementType]]:
    template = GABARIT_TEMPLATES.get(gabarit, GABARIT_TEMPLATES["route"])
    half: list[ElementType] = template["half"]
    center: ElementType | None = template["center"]
    expected_no_center = 2 * len(half)

    if center is not None and n_bands == expected_no_center + 1:
        left = [(Side.GAUCHE, t) for t in half]
        right = [(Side.DROITE, t) for t in reversed(half)]
        return left + [(Side.GAUCHE, center)] + right
    if n_bands == expected_no_center:
        left = [(Side.GAUCHE, t) for t in half]
        right = [(Side.DROITE, t) for t in reversed(half)]
        return left + right

    # Band count doesn't match the chosen template: fall back to a neutral
    # split so nothing is silently mislabeled — the user resolves it via the
    # override dropdown in the mapping UI.
    mid = n_bands // 2
    return [(Side.GAUCHE, ElementType.NON_UTILISE)] * mid + [(Side.DROITE, ElementType.NON_UTILISE)] * (n_bands - mid)


def _frange(start: float, stop: float, step: float):
    n = int(math.floor((stop - start) / step + 1e-9)) + 1
    for i in range(n):
        yield start + i * step


def _extract_heuristic(
    doc,
    state: StateKind,
    axis: AxisReference,
    gabarit: str,
    step_m: float,
    band_overrides: dict[str, tuple[Side, ElementType]],
) -> tuple[list[Band], list[WidthSample]]:
    lines = find_legacy_polylines(doc)
    if len(lines) < 2:
        raise ValueError(
            "Heuristic DXF extraction needs at least 2 legacy POLYLINE boundary lines "
            "(LWPOLYLINE fragments are ignored on purpose, see brief)"
        )
    order = order_lines_by_offset(axis.axis, lines)
    ordered_lines = [PolylineIndex.from_points(lines[i]) for i in order]
    n_bands = len(ordered_lines) - 1
    default_labels = _default_band_labels(n_bands, gabarit)
    stations = list(_frange(0.0, axis.axis.length, step_m))

    bands: list[Band] = []
    samples: list[WidthSample] = []
    for band_idx in range(n_bands):
        band_key = f"dxf-{state.value}-band{band_idx}"
        side, element_type = default_labels[band_idx]
        source, confidence = SourceMethod.RECUPERATION_DXF, 0.4
        if band_key in band_overrides:
            side, element_type = band_overrides[band_key]
            source, confidence = SourceMethod.MENU_DEROULANT, 1.0

        widths: list[float] = []
        for s in stations:
            point, direction = axis.axis.point_and_direction_at_station(s)
            perp = perpendicular_direction(direction)
            p1 = ordered_lines[band_idx].intersect_ray(point, perp)
            p2 = ordered_lines[band_idx + 1].intersect_ray(point, perp)
            if p1 is None or p2 is None:
                continue
            width = math.dist(p1, p2)
            pk = axis.station_to_pk(s)
            widths.append(width)
            samples.append(
                WidthSample(pk=pk, side=side, element_type=element_type, state=state, width_m=width, source=source, band_id=band_key)
            )

        bands.append(
            Band(
                band_id=band_key,
                state=state,
                side=side,
                element_type=element_type,
                source=source,
                confidence=confidence,
                label_hint=f"bande géométrique #{band_idx}",
                sample_count=len(widths),
                width_min=min(widths, default=None),
                width_max=max(widths, default=None),
                width_mean=(sum(widths) / len(widths)) if widths else None,
            )
        )
    return bands, samples


def _parse_layer_tokens(name: str) -> tuple[Side, ElementType]:
    upper = name.upper()
    side = Side.DROITE if re.search(r"(?:^|[-_])D(?:$|[-_])|DROITE|S2", upper) else Side.GAUCHE
    element_type = ElementType.NON_UTILISE
    for pattern, et in _ELEMENT_TOKENS:
        if re.search(pattern, upper):
            element_type = et
            break
    return side, element_type


def _extract_layer_based(
    doc,
    state: StateKind,
    axis: AxisReference,
    band_overrides: dict[str, tuple[Side, ElementType]],
) -> tuple[list[Band], list[WidthSample]]:
    layers = named_axis_and_cote_layers(doc)
    cote_layers = set(layers["cote"])
    texts = read_texts(doc)

    by_layer: dict[str, list] = defaultdict(list)
    for t in texts:
        if t.layer in cote_layers:
            by_layer[t.layer].append(t)

    bands: list[Band] = []
    samples: list[WidthSample] = []
    for layer_name, layer_texts in by_layer.items():
        side, element_type = _parse_layer_tokens(layer_name)
        band_key = f"dxf-{state.value}-{layer_name}"
        source, confidence = SourceMethod.RECUPERATION_ENTREES, (1.0 if element_type != ElementType.NON_UTILISE else 0.2)
        if band_key in band_overrides:
            side, element_type = band_overrides[band_key]
            source, confidence = SourceMethod.MENU_DEROULANT, 1.0

        widths: list[float] = []
        for t in layer_texts:
            value = t.numeric_value
            if value is None:
                continue
            pk, _offset = axis.project((t.x, t.y))
            widths.append(value)
            samples.append(
                WidthSample(pk=pk, side=side, element_type=element_type, state=state, width_m=value, source=source, band_id=band_key)
            )

        bands.append(
            Band(
                band_id=band_key,
                state=state,
                side=side,
                element_type=element_type,
                source=source,
                confidence=confidence,
                label_hint=layer_name,
                sample_count=len(widths),
                width_min=min(widths, default=None),
                width_max=max(widths, default=None),
                width_mean=(sum(widths) / len(widths)) if widths else None,
            )
        )
    return bands, samples

File: frontend/src/engine/dxf/dxfExtractor.ts
/** Port of backend/app/extraction/dxf_extractor.py. */

import type { AxisReference } from "../axisReference";
import { PolylineIndex, orderLinesByOffset, perpendicularDirection, type Point } from "../geometry";
import type { Band, ElementType, Side, SourceMethod, StateKind, WidthSample } from "../../types/domain";
import { parseDxf } from "./dxfReader";
import { findLegacyPolylines, namedAxisAndCoteLayers, numericValue } from "./dxfCommon";
import { maxOf, minOf } from "../arrayUtils";

const ROUTE_TEMPLATE_HALF: ElementType[] = ["accotement", "trottoir", "cycle", "voie"];
const AUTOROUTE_TEMPLATE_HALF: ElementType[] = ["accotement", "bau", "voie", "voie"];

const GABARIT_TEMPLATES: Record<string, { half: ElementType[]; center: ElementType | null }> = {
  route: { half: ROUTE_TEMPLATE_HALF, center: null },
  autoroute: { half: AUTOROUTE_TEMPLATE_HALF, center: "tpc" },
};

const ELEMENT_TOKENS: [RegExp, ElementType][] = [
  [/BAU/i, "bau"],
  [/ACCOT/i, "accotement"],
  [/TR.{0,4}OIR/i, "trottoir"],
  [/CYCL/i, "cycle"],
  [/VOIE|CHAUSS/i, "voie"],
  [/TPC/i, "tpc"],
];

function defaultBandLabels(nBands: number, gabarit: string): [Side, ElementType][] {
  const template = GABARIT_TEMPLATES[gabarit] ?? GABARIT_TEMPLATES.route;
  const { half, center } = template;
  const expectedNoCenter = 2 * half.length;

  if (center !== null && nBands === expectedNoCenter + 1) {
    const left: [Side, ElementType][] = half.map((t) => ["gauche", t]);
    const right: [Side, ElementType][] = [...half].reverse().map((t) => ["droite", t]);
    return [...left, ["gauche", center], ...right];
  }
  if (nBands === expectedNoCenter) {
    const left: [Side, ElementType][] = half.map((t) => ["gauche", t]);
    const right: [Side, ElementType][] = [...half].reverse().map((t) => ["droite", t]);
    return [...left, ...right];
  }
  const mid = Math.floor(nBands / 2);
  const labels: [Side, ElementType][] = [];
  for (let i = 0; i < mid; i++) labels.push(["gauche", "non_utilise"]);
  for (let i = mid; i < nBands; i++) labels.push(["droite", "non_utilise"]);
  return labels;
}

function frange(start: number, stop: number, step: number): number[] {
  const n = Math.floor((stop - start) / step + 1e-9) + 1;
  return Array.from({ length: n }, (_, i) => start + i * step);
}

export function detectDxfMode(content: string): "layers" | "heuristic" {
  const doc = parseDxf(content);
  return namedAxisAndCoteLayers(doc) !== null ? "layers" : "heuristic";
}

export function extractDxfState(
  content: string,
  state: StateKind,
  axis: AxisReference,
  gabarit = "route",
  stepM = 5.0,
): { bands: Band[]; samples: WidthSample[]; mode: "layers" | "heuristic" } {
  const doc = parseDxf(content);
  const mode = namedAxisAndCoteLayers(doc) !== null ? "layers" : "heuristic";

  if (mode === "layers") {
    const { bands, samples } = extractLayerBased(doc, state, axis);
    return { bands, samples, mode };
  }
  const { bands, samples } = extractHeuristic(doc, state, axis, gabarit, stepM);
  return { bands, samples, mode };
}

function extractHeuristic(
  doc: ReturnType<typeof parseDxf>,
  state: StateKind,
  axis: AxisReference,
  gabarit: string,
  stepM: number,
): { bands: Band[]; samples: WidthSample[] } {
  const lines = findLegacyPolylines(doc);
  if (lines.length < 2) {
    throw new Error(
      "L'extraction heuristique DXF nécessite au moins 2 lignes POLYLINE anciennes génération (les LWPOLYLINE fragmentées sont ignorées volontairement)",
    );
  }
  const order = orderLinesByOffset(axis.axis, lines);
  const orderedLines = order.map((i) => new PolylineIndex(lines[i]));
  const nBands = orderedLines.length - 1;
  const defaultLabels = defaultBandLabels(nBands, gabarit);
  const stations = frange(0, axis.axis.length, stepM);

  const bands: Band[] = [];
  const samples: WidthSample[] = [];
  for (let bandIdx = 0; bandIdx < nBands; bandIdx++) {
    const bandKey = `dxf-${state}-band${bandIdx}`;
    const [side, elementType] = defaultLabels[bandIdx];
    const source: SourceMethod = "recuperation_dxf";
    const confidence = 0.4;

    const widths: number[] = [];
    for (const s of stations) {
      const { point, direction } = axis.axis.pointAndDirectionAtStation(s);
      const perp = perpendicularDirection(direction);
      const p1 = orderedLines[bandIdx].intersectRay(point, perp);
      const p2 = orderedLines[bandIdx + 1].intersectRay(point, perp);
      if (!p1 || !p2) continue;
      const width = Math.hypot(p2[0] - p1[0], p2[1] - p1[1]);
      const pk = axis.stationToPk(s);
      widths.push(width);
      samples.push({ pk, side, element_type: elementType, state, width_m: width, source, band_id: bandKey });
    }

    bands.push({
      band_id: bandKey,
      state,
      side,
      element_type: elementType,
      source,
      confidence,
      label_hint: `bande géométrique #${bandIdx}`,
      sample_count: widths.length,
      width_min: widths.length ? minOf(widths) : null,
      width_max: widths.length ? maxOf(widths) : null,
      width_mean: widths.length ? widths.reduce((a, b) => a + b, 0) / widths.length : null,
    });
  }
  return { bands, samples };
}

function parseLayerTokens(name: string): [Side, ElementType] {
  const side: Side = /(?:^|[-_])D(?:$|[-_])|DROITE|S2/i.test(name) ? "droite" : "gauche";
  let elementType: ElementType = "non_utilise";
  for (const [pattern, et] of ELEMENT_TOKENS) {
    if (pattern.test(name)) {
      elementType = et;
      break;
    }
  }
  return [side, elementType];
}

function extractLayerBased(
  doc: ReturnType<typeof parseDxf>,
  state: StateKind,
  axis: AxisReference,
): { bands: Band[]; samples: WidthSample[] } {
  const layers = namedAxisAndCoteLayers(doc);
  const coteLayers = new Set(layers?.cote ?? []);
  const byLayer = new Map<string, typeof doc.texts>();
  for (const t of doc.texts) {
    if (coteLayers.has(t.layer)) {
      if (!byLayer.has(t.layer)) byLayer.set(t.layer, []);
      byLayer.get(t.layer)!.push(t);
    }
  }

  const bands: Band[] = [];
  const samples: WidthSample[] = [];
  for (const [layerName, layerTexts] of byLayer.entries()) {
    const [side, elementType] = parseLayerTokens(layerName);
    const bandKey = `dxf-${state}-${layerName}`;
    const source: SourceMethod = "recuperation_entrees";
    const confidence = elementType !== "non_utilise" ? 1.0 : 0.2;

    const widths: number[] = [];
    for (const t of layerTexts) {
      const value = numericValue(t);
      if (value === null) continue;
      const [pk] = axis.project([t.x, t.y] as Point);
      widths.push(value);
      samples.push({ pk, side, element_type: elementType, state, width_m: value, source, band_id: bandKey });
    }

    bands.push({
      band_id: bandKey,
      state,
      side,
      element_type: elementType,
      source,
      confidence,
      label_hint: layerName,
      sample_count: widths.length,
      width_min: widths.length ? minOf(widths) : null,
      width_max: widths.length ? maxOf(widths) : null,
      width_mean: widths.length ? widths.reduce((a, b) => a + b, 0) / widths.length : null,
    });
  }
  return { bands, samples };
}
